badge_number.go

```diff
@@ -0,0 +1,71 @@
+package apns
+
+import (
+	"errors"
+	"strconv"
+)
+
+// BadgeNumber is much a NullInt64 in
+// database/sql except instead of using
+// the nullable value for driver.Value
+// encoding and decoding, this is specifically
+// meant for JSON encoding and decoding
+type BadgeNumber struct {
+	Number int
+	Valid  bool
+}
+
+// NewBadgeNumber returns a valid BadgeNumber
+// with the value passed in
+func NewBadgeNumber(number int) BadgeNumber {
+	return BadgeNumber{
+		Number: number,
+		Valid:  true,
+	}
+}
+
+// Unset will reset the BadgeNumber to
+// both 0 and invalid
+func (b *BadgeNumber) Unset() {
+	b.Number = 0
+	b.Valid = false
+}
+
+// Set will set the BadgeNumber value to
+// the number passed in, assuming it's >= 0.
+// If so, the BadgeNumber will also be marked valid
+func (b *BadgeNumber) Set(number int) error {
+	if number < 0 {
+		return errors.New("Number must be >= 0")
+	}
+
+	b.Number = number
+	b.Valid = true
+	return nil
+}
+
```

Comment:
If we make this method take a [`uint`](https://golang.org/ref/spec#Numeric_types) we can get away without the `error` return type (and make this method safe at compile time).